File: eclipse-plugin/src/main/java/xyz/devglobe/eclipse/auth/ConfigWriter.java
package xyz.devglobe.eclipse.auth;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.attribute.PosixFilePermission;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import xyz.devglobe.eclipse.core.DevGlobePlugin;

/**
 * Reads and writes the DevGlobe config.toml file (~/.devglobe/config.toml).
 * The API key is written directly into the TOML root so setup works before
 * the core binary has been downloaded, and the key never appears on a process
 * command line.
 */
public final class ConfigWriter {

    private static final Pattern ROOT_API_KEY = Pattern.compile("^api_key\\s*=\\s*\"([^\"]*)\"");

    private ConfigWriter() {}

    public static File devglobeDir() {
        return new File(System.getProperty("user.home"), ".devglobe");
    }

    public static File configPath() {
        return new File(devglobeDir(), "config.toml");
    }

    public static File logPath() {
        return new File(devglobeDir(), "devglobe.log");
    }

    // ── API Key ──────────────────────────────────────────────────────────

    /**
     * Writes the API key into the {@code api_key} root entry of config.toml,
     * preserving any other settings and sections. Written with 0600 permissions.
     */
    public static void writeApiKey(String apiKey) {
        try {
            File dir = devglobeDir();
            if (!dir.exists()) dir.mkdirs();

            File configFile = configPath();
            List<String> lines = configFile.exists()
                    ? new ArrayList<>(Files.readAllLines(configFile.toPath()))
                    : new ArrayList<>();

            String keyLine = "api_key = \"" + escapeToml(apiKey) + "\"";
            int idx = findRootApiKeyIndex(lines);
            if (idx >= 0) {
                lines.set(idx, keyLine);
            } else {
                lines.add(0, keyLine);
            }

            String output = String.join("\n", lines).replaceAll("\n{3,}", "\n\n");
            if (!output.endsWith("\n")) output += "\n";
            Files.writeString(configFile.toPath(), output);
            setRestrictivePermissions(configFile);
        } catch (IOException e) {
            DevGlobePlugin.log("Failed to write API key: " + e.getMessage());
        }
    }

    private static String escapeToml(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    public static boolean hasApiKey() {
        try {
            File configFile = configPath();
            if (!configFile.exists()) return false;
            List<String> lines = Files.readAllLines(configFile.toPath());
            int idx = findRootApiKeyIndex(lines);
            if (idx < 0) return false;
            Matcher m = ROOT_API_KEY.matcher(lines.get(idx).trim());
            return m.matches() && !m.group(1).isEmpty();
        } catch (IOException e) {
            return false;
        }
    }

    public static String readApiKey() {
        try {
            File configFile = configPath();
            if (!configFile.exists()) return null;
            List<String> lines = Files.readAllLines(configFile.toPath());
            int idx = findRootApiKeyIndex(lines);
            if (idx < 0) return null;
            Matcher m = ROOT_API_KEY.matcher(lines.get(idx).trim());
            return m.matches() ? m.group(1) : null;
        } catch (IOException e) {
            return null;
        }
    }

    /**
     * Clears the API key by removing the {@code api_key} line from config.toml.
     * This avoids calling {@code devglobe-core setup ""} which doesn't accept
     * an empty argument.
     */
    public static void clearApiKey() {
        try {
            File configFile = configPath();
            if (!configFile.exists()) return;

            List<String> lines = new ArrayList<>(Files.readAllLines(configFile.toPath()));
            int idx = findRootApiKeyIndex(lines);
            if (idx < 0) return; // no key to clear

            lines.remove(idx);
            String output = String.join("\n", lines).replaceAll("\n{3,}", "\n\n");
            if (!output.endsWith("\n")) output += "\n";
            Files.writeString(configFile.toPath(), output);
            setRestrictivePermissions(configFile);
        } catch (IOException e) {
            DevGlobePlugin.log("Failed to clear API key: " + e.getMessage());
        }
    }

    // ── Debug ────────────────────────────────────────────────────────────

    private static final Pattern DEBUG_PATTERN = Pattern.compile("^debug\\s*=\\s*(true|false)");

    public static boolean isDebugEnabled() {
        try {
            File configFile = configPath();
            if (!configFile.exists()) return false;
            for (String line : Files.readAllLines(configFile.toPath())) {
                String trimmed = line.trim();
                if (trimmed.startsWith("[")) return false;
                Matcher m = DEBUG_PATTERN.matcher(trimmed);
                if (m.matches()) return Boolean.parseBoolean(m.group(1));
            }
            return false;
        } catch (IOException e) {
            return false;
        }
    }

    public static void setDebug(boolean enabled) {
        try {
            File dir = devglobeDir();
            if (!dir.exists()) dir.mkdirs();

            File configFile = configPath();
            List<String> lines = configFile.exists()
                    ? new ArrayList<>(Files.readAllLines(configFile.toPath()))
                    : new ArrayList<>();

            int idx = -1;
            for (int i = 0; i < lines.size(); i++) {
                String trimmed = lines.get(i).trim();
                if (trimmed.startsWith("[")) break;
                if (trimmed.startsWith("debug")) { idx = i; break; }
            }

            if (idx >= 0 && enabled) {
                lines.set(idx, "debug = true");
            } else if (idx >= 0 && !enabled) {
                lines.remove(idx);
            } else if (idx < 0 && enabled) {
                int apiKeyIdx = -1;
                for (int i = 0; i < lines.size(); i++) {
                    if (lines.get(i).trim().startsWith("api_key")) { apiKeyIdx = i; break; }
                }
                if (apiKeyIdx >= 0) lines.add(apiKeyIdx + 1, "debug = true");
                else lines.add(0, "debug = true");
            }

            String output = String.join("\n", lines).replaceAll("\n{3,}", "\n\n");
            if (!output.endsWith("\n")) output += "\n";
            Files.writeString(configFile.toPath(), output);
            setRestrictivePermissions(configFile);
        } catch (IOException e) {
            DevGlobePlugin.log("Failed to set debug flag: " + e.getMessage());
        }
    }

    // ── Helpers ──────────────────────────────────────────────────────────

    private static int findRootApiKeyIndex(List<String> lines) {
        for (int i = 0; i < lines.size(); i++) {
            String trimmed = lines.get(i).trim();
            if (trimmed.startsWith("[")) return -1;
            if (trimmed.startsWith("api_key")) return i;
        }
        return -1;
    }

    private static void setRestrictivePermissions(File file) {
        try {
            String os = System.getProperty("os.name", "").toLowerCase();
            if (os.contains("win")) return;
            Set<PosixFilePermission> perms = EnumSet.of(
                    PosixFilePermission.OWNER_READ,
                    PosixFilePermission.OWNER_WRITE);
            Files.setPosixFilePermissions(file.toPath(), perms);
        } catch (Exception ignored) {
            // best effort
        }
    }
}

File: jetbrains-plugin/src/main/kotlin/xyz/devglobe/plugin/auth/ConfigWriter.kt
package xyz.devglobe.plugin.auth

import com.intellij.openapi.diagnostic.Logger
import java.io.File
import java.nio.file.attribute.PosixFilePermissions
import java.nio.file.Files

/**
 * Writes the API key to ~/.devglobe/config.toml so the core (running as a
 * subprocess) can pick it up on next init. Preserves other settings.
 */
object ConfigWriter {

    private val LOG = Logger.getInstance(ConfigWriter::class.java)

    private fun devglobeDir(): File = File(System.getProperty("user.home"), ".devglobe")
    fun configPath(): File = File(devglobeDir(), "config.toml")
    fun logPath(): File = File(devglobeDir(), "devglobe.log")

    fun writeApiKey(apiKey: String) {
        try {
            val dir = devglobeDir()
            if (!dir.exists()) dir.mkdirs()

            val configFile = configPath()
            val original = if (configFile.exists()) configFile.readText().split("\n") else emptyList()
            val newKeyLine = "api_key = \"$apiKey\""

            val lines = original.toMutableList()
            val existingIdx = findRootApiKeyIndex(lines)
            if (existingIdx >= 0) {
                lines[existingIdx] = newKeyLine
            } else {
                lines.add(0, newKeyLine)
            }

            val output = lines.joinToString("\n").replace(Regex("\n{3,}"), "\n\n")
            configFile.writeText(if (output.endsWith("\n")) output else "$output\n")
            setRestrictivePermissions(configFile)
        } catch (e: Exception) {
            LOG.warn("Failed to write config.toml: ${e.message}")
        }
    }

    fun hasApiKey(): Boolean {
        return try {
            val configFile = configPath()
            if (!configFile.exists()) return false
            val lines = configFile.readText().split("\n")
            val idx = findRootApiKeyIndex(lines)
            idx >= 0 && lines[idx].substringAfter("=").trim().trim('"').isNotEmpty()
        } catch (e: Exception) {
            LOG.warn("Failed to read config.toml: ${e.message}")
            false
        }
    }

    fun isDebugEnabled(): Boolean {
        return try {
            val configFile = configPath()
            if (!configFile.exists()) return false
            for (line in configFile.readText().split("\n")) {
                val trimmed = line.trim()
                if (trimmed.startsWith("[")) return false
                val m = Regex("^debug\\s*=\\s*(true|false)").find(trimmed) ?: continue
                return m.groupValues[1] == "true"
            }
            false
        } catch (e: Exception) {
            LOG.warn("Failed to read debug flag: ${e.message}")
            false
        }
    }

    fun setDebug(enabled: Boolean) {
        try {
            val dir = devglobeDir()
            if (!dir.exists()) dir.mkdirs()

            val configFile = configPath()
            val original = if (configFile.exists()) configFile.readText().split("\n") else emptyList()
            val lines = original.toMutableList()

            // Find or remove existing root-level debug line.
            var idx = -1
            for ((i, line) in lines.withIndex()) {
                val trimmed = line.trim()
                if (trimmed.startsWith("[")) break
                if (trimmed.startsWith("debug")) { idx = i; break }
            }

            when {
                idx >= 0 && enabled -> lines[idx] = "debug = true"
                idx >= 0 && !enabled -> lines.removeAt(idx)
                idx < 0 && enabled -> {
                    val apiKeyIdx = lines.indexOfFirst { it.trim().startsWith("api_key") }
                    if (apiKeyIdx >= 0) lines.add(apiKeyIdx + 1, "debug = true")
                    else lines.add(0, "debug = true")
                }
            }

            val output = lines.joinToString("\n").replace(Regex("\n{3,}"), "\n\n")
            configFile.writeText(if (output.endsWith("\n")) output else "$output\n")
            setRestrictivePermissions(configFile)
        } catch (e: Exception) {
            LOG.warn("Failed to set debug flag: ${e.message}")
        }
    }

    fun clearApiKey() {
        try {
            val configFile = configPath()
            if (!configFile.exists()) return
            val kept = mutableListOf<String>()
            var beforeSection = true
            for (line in configFile.readText().split("\n")) {
                val trimmed = line.trim()
                if (trimmed.startsWith("[")) beforeSection = false
                if (beforeSection && trimmed.startsWith("api_key")) continue
                kept.add(line)
            }
            configFile.writeText(kept.joinToString("\n"))
            setRestrictivePermissions(configFile)
        } catch (e: Exception) {
            LOG.warn("Failed to clear api_key from config.toml: ${e.message}")
        }
    }

    // Returns the index of the root-level `api_key = ...` line, or -1.
    // Only scans lines before the first `[section]` since api_key is in the
    // implicit root table.
    private fun findRootApiKeyIndex(lines: List<String>): Int {
        for ((i, line) in lines.withIndex()) {
            val trimmed = line.trim()
            if (trimmed.startsWith("[")) return -1
            if (trimmed.startsWith("api_key")) return i
        }
        return -1
    }

    private fun setRestrictivePermissions(file: File) {
        try {
            if (System.getProperty("os.name").lowercase().contains("win")) return
            Files.setPosixFilePermissions(file.toPath(), PosixFilePermissions.fromString("rw-------"))
        } catch (_: Exception) {
            // best effort; not critical
        }
    }
}
